File: bitboard.py
# Define pieces for white and black
WHITE_PAWNS   = ((1 << 8) - 1) << 8
WHITE_KNIGHTS = (1 << 1) | (1 << 6)
WHITE_BISHOPS = (1 << 2) | (1 << 5)
WHITE_ROOKS   = (1 << 0) | (1 << 7)
WHITE_QUEEN   = (1 << 3)
WHITE_KING    = (1 << 4)

BLACK_PAWNS   = ((1 << 8) - 1) << 48
BLACK_KNIGHTS = (1 << 57) | (1 << 62)
BLACK_BISHOPS = (1 << 58) | (1 << 61)
BLACK_ROOKS   = (1 << 56) | (1 << 63)
BLACK_QUEEN   = (1 << 59)
BLACK_KING    = (1 << 60)

WHITE_PIECES = WHITE_PAWNS | WHITE_BISHOPS | WHITE_KING | \
               WHITE_KNIGHTS | WHITE_QUEEN | WHITE_ROOKS

BLACK_PIECES = BLACK_PAWNS | BLACK_BISHOPS | BLACK_KING | \
               BLACK_KNIGHTS | BLACK_QUEEN | BLACK_ROOKS

# Define rank and files for chessboard
RANK_1 = 0x00000000000000FF
RANK_2 = RANK_1 << 8
RANK_3 = RANK_2 << 8
RANK_4 = RANK_3 << 8
RANK_5 = RANK_4 << 8
RANK_6 = RANK_5 << 8
RANK_7 = RANK_6 << 8
RANK_8 = RANK_7 << 8

FILE_A = 0x0101010101010101
FILE_B = FILE_A << 1
FILE_C = FILE_B << 2
FILE_D = FILE_C << 3
FILE_E = FILE_D << 4
FILE_F = FILE_E << 5
FILE_G = FILE_F << 6
FILE_H = FILE_G << 7

# Generate pre-computed squares
SQUARES = [1 << square for square in range(64)]

# Define actions that a player can take
class Move():
    def __init__(self, square=None, is_resign=False):
        assert (square is not None) ^ is_resign
        self.square = square
        self.play = self.square is not None
        self.is_resign = is_resign

    @classmethod
    def play(cls, square):
        return Move(square=square)
    
    @classmethod
    def is_resign(cls):
        return Move(is_resign=True)

# Print a bitboard
def show_bitboard(bitboard):
    print("  +-----------------+")

    for row in range(7, -1, -1):
        print(f"{row + 1} |", end=" ")
        for col in range(8):
            square = row * 8 + col
            if bitboard & SQUARES[square]:
                print(1, end=" ")
            else:
                print(".", end=" ")
        print(f"|", end=" ")
        print()

    print("  +-----------------+")
    print("    a b c d e f g h")
    print()

# Set a piece on the board
def set_piece(bitboard, square):
    return bitboard | SQUARES[square]

# Clear the piece from the board if it is occupied
def clear_piece(bitboard, square):
    return bitboard & ~SQUARES[square]
# Toggle a piece
def toggle_piece(bitboard, square):
    return bitboard ^ SQUARES[square]

# Check if square is occupied
def is_occupied(bitboard, square):
    return bitboard & SQUARES[square]

# Function to get all occupied squares
def occupied_squares():
    return (WHITE_PIECES | BLACK_PIECES)

# Find LSB in a bitboard
def get_LSB(bitboard):
    if bitboard == 0:
        return -1
    return (bitboard & -bitboard).bit_length() - 1

# Find MSB in a bitboard
def get_MSB(bitboard):
    if bitboard == 0:
        return -1
    return bitboard.bit_length() - 1

# Define population count (check no of pieces)
def popcount(bitboard):
    return bin(bitboard).count("1")

# Flip board vertically
def flip_vertical(bitboard):
    k1 = 0x00FF00FF00FF00FF
    k2 = 0x0000FFFF0000FFFF

    bitboard = ((bitboard >>  8) & k1) | ((bitboard & k1) <<  8)
    bitboard = ((bitboard >> 16) & k2) | ((bitboard & k2) << 16)
    bitboard = (bitboard >> 32) | (bitboard << 32)

    return bitboard

# Mirror board horizontally
def mirror_horizontal(bitboard):
    k1 = 0x5555555555555555
    k2 = 0x3333333333333333
    k4 = 0x0F0F0F0F0F0F0F0F

    bitboard = ((bitboard >> 1) & k1) | ((bitboard & k1) << 1)
    bitboard = ((bitboard >> 2) & k2) | ((bitboard & k2) << 2)
    bitboard = ((bitboard >> 4) & k4) | ((bitboard & k4) << 4)

    return bitboard

# Flip board along A1-H8 diagonal
def flip_diagonal(bitboard):
    k1 = 0x5500550055005500
    k2 = 0x3333000033330000
    k4 = 0x0f0f0f0f00000000

    t = k4 & (bitboard ^ (bitboard << 28))
    bitboard ^= t ^ (t >> 28)
    t = k2 & (bitboard ^ (bitboard << 14))
    bitboard ^= t ^ (t >> 14)
    t = k1 & (bitboard ^ (bitboard << 7))
    bitboard ^= t ^ (t >> 7)

    return bitboard

# Flip board along A8-H1 diagonal
def flip_anti_diagonal(bitboard):
    k1 = 0xaa00aa00aa00aa00
    k2 = 0xcccc0000cccc0000
    k4 = 0xf0f0f0f00f0f0f0f

    t = bitboard ^ (bitboard << 36)
    bitboard ^= k4 & (t ^ (bitboard >> 36))
    t = k2 & (bitboard ^ (bitboard << 18))
    bitboard ^= t ^ (t >> 18)
    t = k1 & (bitboard ^ (bitboard << 9))
    bitboard ^= t ^ (t >> 9)

    return bitboard

# Functions for moving one step in any direction
def north(bitboard):
    return bitboard << 8

def south(bitboard):
    return bitboard >> 8

def north_east(bitboard):
    return (bitboard << 9) & ~FILE_A

def south_east(bitboard):
    return (bitboard >> 7) & ~FILE_A

def north_west(bitboard):
    return (bitboard << 7) & ~FILE_H

def south_west(bitboard):
    return (bitboard >> 9) & ~FILE_H

def east(bitboard):
    return (bitboard << 1) & ~FILE_A

def west(bitboard):
    return (bitboard >> 1) & ~FILE_H

File: bitboard.cpp
#include "bitboard.hpp"
#include <iostream>
#include <bitset>
#include <vector>
#include <algorithm>
#include <cstdint>

// Define pieces for white and black
constexpr uint64_t WHITE_PAWNS   = ((1ULL << 8) - 1) << 8;
constexpr uint64_t WHITE_KNIGHTS = (1ULL << 1) | (1ULL << 6);
constexpr uint64_t WHITE_BISHOPS = (1ULL << 2) | (1ULL << 5);
constexpr uint64_t WHITE_ROOKS   = (1ULL << 0) | (1ULL << 7);
constexpr uint64_t WHITE_QUEEN   = (1ULL << 3);
constexpr uint64_t WHITE_KING    = (1ULL << 4);

constexpr uint64_t BLACK_PAWNS   = ((1ULL << 8) - 1) << 48;
constexpr uint64_t BLACK_KNIGHTS = (1ULL << 57) | (1ULL << 62);
constexpr uint64_t BLACK_BISHOPS = (1ULL << 58) | (1ULL << 61);
constexpr uint64_t BLACK_ROOKS   = (1ULL << 56) | (1ULL << 63);
constexpr uint64_t BLACK_QUEEN   = (1ULL << 59);
constexpr uint64_t BLACK_KING    = (1ULL << 60);

constexpr uint64_t WHITE_PIECES = WHITE_PAWNS | WHITE_BISHOPS | WHITE_KING |
                                   WHITE_KNIGHTS | WHITE_QUEEN | WHITE_ROOKS;

constexpr uint64_t BLACK_PIECES = BLACK_PAWNS | BLACK_BISHOPS | BLACK_KING |
                                   BLACK_KNIGHTS | BLACK_QUEEN | BLACK_ROOKS;

// Define rank and files for chessboard
constexpr uint64_t RANK_1 = 0x00000000000000FF;
constexpr uint64_t RANK_2 = RANK_1 << 8;
constexpr uint64_t RANK_3 = RANK_2 << 8;
constexpr uint64_t RANK_4 = RANK_3 << 8;
constexpr uint64_t RANK_5 = RANK_4 << 8;
constexpr uint64_t RANK_6 = RANK_5 << 8;
constexpr uint64_t RANK_7 = RANK_6 << 8;
constexpr uint64_t RANK_8 = RANK_7 << 8;

constexpr uint64_t FILE_A = 0x0101010101010101;
constexpr uint64_t FILE_B = FILE_A << 1;
constexpr uint64_t FILE_C = FILE_B << 1;
constexpr uint64_t FILE_D = FILE_C << 1;
constexpr uint64_t FILE_E = FILE_D << 1;
constexpr uint64_t FILE_F = FILE_E << 1;
constexpr uint64_t FILE_G = FILE_F << 1;
constexpr uint64_t FILE_H = FILE_G << 1;

constexpr uint64_t NOT_A_FILE = ~FILE_A;
constexpr uint64_t NOT_H_FILE = ~FILE_H;
constexpr uint64_t NOT_AB_FILE = ~FILE_A & ~FILE_B;
constexpr uint64_t NOT_HG_FILE = ~FILE_H & ~FILE_G;

// Define co-ordinate map
std::vector<std::string> coord = {
    "a1", "b1", "c1", "d1", "e1", "f1", "g1", "h1",
    "a2", "b2", "c2", "d2", "e2", "f2", "g2", "h2",
    "a3", "b3", "c3", "d3", "e3", "f3", "g3", "h3",
    "a4", "b4", "c4", "d4", "e4", "f4", "g4", "h4",
    "a5", "b5", "c5", "d5", "e5", "f5", "g5", "h5",
    "a6", "b6", "c6", "d6", "e6", "f6", "g6", "h6",
    "a7", "b7", "c7", "d7", "e7", "f7", "g7", "h7",
    "a8", "b8", "c8", "d8", "e8", "f8", "g8", "h8"
};

uint64_t coordToBitboard(const std::string& square) {
    // Search for the square in the `coord` vector
    auto it = std::find(coord.begin(), coord.end(), square);

    // If the square is not found, throw an exception
    if (it == coord.end()) {
        throw std::invalid_argument("Invalid coordinate: " + square);
    }

    // Calculate the index of the square and return the corresponding bitboard
    int index = std::distance(coord.begin(), it);
    return 1ULL << index;
}

// Precomputed squares
std::vector<uint64_t> initSquares() {
    std::vector<uint64_t> squares(64);
    for (int i = 0; i < 64; ++i) {
        squares[i] = 1ULL << i;
    }
    return squares;
}
std::vector<uint64_t> SQUARES = initSquares();

// Display a bitboard
void showBitboard(uint64_t bb) {
    std::cout << "  +-----------------+\n";
    for (int row = 7; row >= 0; --row) {
        std::cout << row + 1 << " | ";
        for (int col = 0; col < 8; ++col) {
            int square = row * 8 + col;
            if (bb & SQUARES[square]) {
                std::cout << "1 ";
            } else {
                std::cout << ". ";
            }
        }
        std::cout << "|\n";
    }
    std::cout << "  +-----------------+\n";
    std::cout << "    a b c d e f g h\n\n";
}

// Set a piece on the board
uint64_t setPiece(uint64_t bb, int square) {
    return bb | SQUARES[square];
}

// Clear a piece from the board
uint64_t clearPiece(uint64_t bb, int square) {
    return bb & ~SQUARES[square];
}

// Toggle a piece on the board
uint64_t togglePiece(uint64_t bb, int square) {
    return bb ^ SQUARES[square];
}

// Function to get all occupied squares
uint64_t occupiedSquares() {
    return WHITE_PIECES | BLACK_PIECES;
}

// Check if a square is occupied
bool isOccupied(uint64_t bb, int square) {
    return bb & SQUARES[square];
}

// Get least significant bit (LSB)
int getLSB(uint64_t bb) {
    if (bb == 0) return -1;
    return __builtin_ctzll(bb);
}

// Get most significant bit (MSB)
int getMSB(uint64_t bb) {
    if (bb == 0) return -1;
    return 63 - __builtin_clzll(bb);
}

// Population count (number of set bits)
int popcount(uint64_t bb) {
    return __builtin_popcountll(bb);
}

// Flip board vertically
uint64_t flip_vertical(uint64_t bb) {
    const uint64_t k1 = 0x00FF00FF00FF00FFULL;
    const uint64_t k2 = 0x0000FFFF0000FFFFULL;

    bb = ((bb >> 8) & k1) | ((bb & k1) << 8);
    bb = ((bb >> 16) & k2) | ((bb & k2) << 16);
    bb = (bb >> 32) | (bb << 32);

    return bb;
}

// Mirror board horizontally
uint64_t mirror_horizontal(uint64_t bb) {
    const uint64_t k1 = 0x5555555555555555ULL;
    const uint64_t k2 = 0x3333333333333333ULL;
    const uint64_t k4 = 0x0F0F0F0F0F0F0F0FULL;

    bb = ((bb >> 1) & k1) | ((bb & k1) << 1);
    bb = ((bb >> 2) & k2) | ((bb & k2) << 2);
    bb = ((bb >> 4) & k4) | ((bb & k4) << 4);

    return bb;
}

// Flip board along A1-H8 diagonal
uint64_t flip_diagonal(uint64_t bb) {
    const uint64_t k1 = 0x5500550055005500ULL;
    const uint64_t k2 = 0x3333000033330000ULL;
    const uint64_t k4 = 0x0f0f0f0f00000000ULL;

    uint64_t t;
    t = k4 & (bb ^ (bb << 28));
    bb ^= t ^ (t >> 28);
    t = k2 & (bb ^ (bb << 14));
    bb ^= t ^ (t >> 14);
    t = k1 & (bb ^ (bb << 7));
    bb ^= t ^ (t >> 7);

    return bb;
}

// Flip board along A8-H1 diagonal
uint64_t flip_anti_diagonal(uint64_t bb) {
    const uint64_t k1 = 0xaa00aa00aa00aa00ULL;
    const uint64_t k2 = 0xcccc0000cccc0000ULL;
    const uint64_t k4 = 0xf0f0f0f00f0f0f0fULL;

    uint64_t t;
    t = bb ^ (bb << 36);
    bb ^= k4 & (t ^ (bb >> 36));
    t = k2 & (bb ^ (bb << 18));
    bb ^= t ^ (t >> 18);
    t = k1 & (bb ^ (bb << 9));
    bb ^= t ^ (t >> 9);

    return bb;
}

// Directions for piece movement
uint64_t north(uint64_t bb) {
    return bb << 8;
}

uint64_t south(uint64_t bb) {
    return bb >> 8;
}

uint64_t east(uint64_t bb) {
    return (bb << 1) & NOT_A_FILE;
}

uint64_t west(uint64_t bb) {
    return (bb >> 1) & NOT_H_FILE;
}

uint64_t northEast(uint64_t bb) {
    return (bb << 9) & NOT_A_FILE;
}

uint64_t northWest(uint64_t bb) {
    return (bb << 7) & NOT_H_FILE;
}

uint64_t southEast(uint64_t bb) {
    return (bb >> 7) & NOT_A_FILE;
}

uint64_t southWest(uint64_t bb) {
    return (bb >> 9) & NOT_H_FILE;
}

uint64_t northNE(uint64_t bb) {
    return (bb << 17) & NOT_A_FILE;
}

uint64_t northEE(uint64_t bb) {
    return (bb << 10) & NOT_AB_FILE;
}

uint64_t northNW(uint64_t bb) {
    return (bb << 15) & NOT_H_FILE;
}

uint64_t northWW(uint64_t bb) {
    return (bb << 6) & NOT_HG_FILE;
}

uint64_t southEE(uint64_t bb) {
    return (bb >> 6) & NOT_AB_FILE;
}

uint64_t southSE(uint64_t bb) {
    return (bb >> 15) & NOT_A_FILE;
}

uint64_t southWW(uint64_t bb) {
    return (bb >> 10) & NOT_HG_FILE;
}

uint64_t southSW(uint64_t bb) {
    return (bb >> 17) & NOT_H_FILE;
}